designs/0035-pin-parameters.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: pin-parameters
+- Start Date: 2025-05-05
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+It is often useful to be able to 'pin' the otherwise free parameters in a statistical model to specific values. First, this is useful when debugging a statistical model to diagnose computational problems or understand the priors. Second, it is useful when one wishes to explore a simpler model that is nested inside an extended one. e.g., the mu = 0 no effect model that is nested inside the   mu != 0 model of a new effect of size mu. This proposal makes pinning straight-forward at runtime.
+
+# Motivation
+[motivation]: #motivation
+
+At present, to pin a parameter a Stan model must be rewritten. We must either: 
+
+- move a parameter from the parameter block to the data block, where it is pinned to a fixed value
+- add convoluted logic so that a boolean (more precisely an integer<lower=0, upper=1>) in the data block can control whether a parameter is pinned, e.g., (from [here](https://discourse.mc-stan.org/t/fixing-parameters-in-a-model/39035/4?u=andrewfowlie))
+```
+data {
+  int<lower=0, upper=1> mu_is_data;
+  array[mu_is_data] mu_val;
+  ...
+parameters {
+  array[1 - mu_is_data] mu_param;
+  ...
+transformed parameters {
+  real mu = mu_is data ? mu_val[1] : mu_param[1];
+  ...
+```
+
+These are inelegant and unsatisfactory, partly as they require rewriting and recompiling a model.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new `pin` keyword-value pair in `cmdstan` command-line options that would appear [here](https://mc-stan.org/docs/cmdstan-guide/command_line_options.html). We don't at this stage make any proposal for how this feature would be propagated to other Stan interfaces, they do not anticipate any difficulties. 
+
+## Command-Line Interface Overview
+
+...
+
+- `pin` - specifies values for any parameters that should be pinned, if any
+
+...
+
+### Pin model parameters argument
```

Comment:
This section needs to make it clear that variables are going to be pinned or not pinned---they can't be partially pinned.  Suppose I have the following program.

```stan
vector[10] a;
```

I can pin all 10 components of `a` or none of them.  I can't just pin `a[3]` and `a[7]` and leave the others as parameters, for example.

Reply:
It's also worth noting that this restriction is primarily due to the lack of a good design, NOT due to mathematical concerns like the restriction on constrained parameters.

If we could agree on what the specification for the JSON file that only specifies `a[3]` and `a[7]` looks like, we could plumb enough information through the compiler and inference to make it work